### ISSUE TITLE
[UX] Using star as a wildcard in layout paths does not have expected behavior

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1602]
bug report from email: 
I have result pages where I need the full page width for content. Created custom layout, one column, picked the url string (results/*) as a page identifier, saved. Cleared all the caches. But the standard layout is still used on those pages too. Is it me? Is it the default theme? Hard to guess.

**Comment by jenlampton** [comment]
Most users would expect that entering a path with an asterisk into the layout path field would function as though a second "default" layout were created that applied to all pages under that path. Is there any way we can make this happen automatically when someone enters an asterisk?

**Comment by Graham-72** [comment]
I am wondering whether it may be necessary to also positively exclude results/* on the standard layout? I feel sure I have done something similar and will check.

**Comment by Graham-72** [comment]
Ah no, of course, you have to use % not * as the wildcard.

**Comment by maxmeno** [comment]
My fault - email report was wrong, I actually used % as a wildcard, it doesn't work.

**Comment by jenlampton** [comment]
I think part of the problem is that people expect the % (or star) to work like a wildcard, and stand-in for "anything here".  But that's not how it works.  Instead, the % is a placeholder only for a context. So `user/%` will only work when the % is a user ID, so `user/1` and `user/3` will match, but not `user/login`.

**Comment by jenlampton** [comment]
Here's a use-case that was reported via e-mail as having trouble:

> I'm trying to set-up my site to use a specific "layout" for these paths:
user/login
user/[UID]/edit
user/[UID]/content

**Comment by jenlampton** [comment]
I think this sounds like a shortcoming in layouts, so I have created https://github.com/backdrop/backdrop-issues/issues/1671

**Comment by JugglingCoder** [comment]
"Here's a use-case that was reported via e-mail as having trouble:"
That was me, thank you Jen for posting this. Yes, the only "work-around" I could come up with was "add Layout" (and use the same layout) for each path.  Optimal would be of course: "user/*" and all paths under "user" would use that layout.

But... I'm still confused as to why there is also a "path" in "Add visibility condition"... ??
If the "path" is "user/*" (or "user/%") why would I need a 2nd "path" in visibility conditions?

**Comment by klonos** [comment]
> I think the only current solution for this is to have three different layouts, ...Needing multiple layouts for multiple paths sounds like a shortcoming in layouts, so I have created #1671

:+1: I would have done it if you haven't beaten me to it.

**Comment by docwilmot** [comment]
I've started a solution for this and here documenting my thoughts:

Background
When you visit a page Backdrop checks the path in `layout_get_layout_by_path()`, to see if there is a layout matching **exactly** that path. 

If you visit `company/dept`, then `layout_get_layout_by_path()` will be passed exactly that as the path, so a layout with path `company/%` will not match. 

But  entity load paths (such as `node/3` or `user/5`) are passed to `layout_get_layout_by_path()` *with the wildcard*, so when you visit `node/3`, the function gets `node/%` as the path, so if there is a layout with path matching `node/%` that layout will match.

So what this PR attempts to do is to breakdown a path passed to `layout_get_layout_by_path()` into possible wildcard options, then check to see if there is a layout matching this option. So for a path `company/dept/secretaries/sheila`, we'd check for layouts with a path
- `company/%/secretaries/sheila`
- `company/dept/%/sheila`
- `company/%/secretaries/%`
- `company/dept/%/sheila`
- `company/%/%`
- `company/%`
etc

So if a person visits path `company/dept/secretaries/sheila`: 
- if there is a layout: `company/%` this path should match that layout
- if the same site has a layout `company/%/secretaries` this should **not** match
- if the same site has another layout `company/%/secretaries/%` this should match
- if the same site has a layout `company/%/%` this should match

Need to do more on the cascading of these paths. 

Also, related, a path `node/%/edit` is not matching a layout `node/%/%` because Backdrop is finding both the Admin default and the `node/%/%` layout as designed, but the Admin layout has a higher weight so its chosen first. 

Would be good to consider as well matching by aliases, but perhaps another PR.

**Comment by jenlampton** [comment]
I think you may be confusing a `*` with a `%`. A star is a wildcard and means "anything" but a `%` is a system-placeholder and means that there is a known backdrop system menu router that matches that pattern exactly.

I suspect that *everyone* is likely confusing these two because in the Layout UI the description text on that field actually says that the `%` is a wildcard, when it is not.

I don't think that we should expand the meaning of of `%` to also include the meaning of `*`. I think that's going to confuse people even more. I also don't think the right approach here is to try to explain to people what the percent actually means. 

I'd prefer to re-do that field to use a star instead of the percent, and then do all the expected checking, as explained in the previous comment (but with `*` in place of `%`)

**Comment by docwilmot** [comment]
Not so sure about this. Fact is everyone may be "confused" because a % symbol seems intuitive for this purpose. I think conversely adding a new cryptic symbol is more confusing.

Right now to the end user `node/%/%` means "a path that starts with 'node' followed by any two things". I think that's a benign assumption. Backdrop can simply handle the context stuff in the background.

Would welcome more opinions.

**Comment by jenlampton** [comment]
> I think conversely adding a new cryptic symbol is more confusing.

Yes! This is exactly my point :) 

We added a `%` when Layouts went in. Switching to this new cryptic symbol confused everyone, because what they were already familiar with was a `*`. Now everyone thinks the `%` should behave like a `*` and are confused why it doesn't. 

The proposed suggestion above changes how it behaves so that it would behave more like a `*` but without actually replacing the new cryptic symbol (`%`) with the old cryptic symbol that people already understand (`*`). 

I think this will make things even more confusing if we have two different symbols that do the same thing.  If we do this, people will wonder why a wildcard is represented by a star everywhere except this one Layouts interface, where for some reason it is a `%`.

**Comment by docwilmot** [comment]
I understand all of this, but point is we already have the %.
For a path `node/NID/ANYTHING` what is your suggestion that we adopt to enable this functionality in Backdrop *going forward*? 
1. `node/%/%`
2. `node/%/*`
3. `node/*/*`

**Comment by jenlampton** [comment]
> For a path `node/NID/ANYTHING what is your suggestion that we adopt to enable this functionality in Backdrop going` forward?

Any path that has an `ANYTHING` needs to use a star, so my preference would be `node/*/*`, or even `node/*` would work.

Is this possible right now? IIRC we can only do the pre-defined paths at `node/%/%` (which does not handle the `ANYTHING` part -- the %s would need to be one of the strings Backdrop already knows about)

**Comment by docwilmot** [comment]
So are you proposing we change the symbol used in layouts totally to an asterisk?

**Comment by jenlampton** [comment]
Yes. If we're going to change it to work like an asterisk (which I think we should -- that's what people expect) then it should be an asterisk.

**Comment by docwilmot** [comment]
Interesting: seems @jenlampton makes more sense after a 9-month break and two cups of coffee: I agree wholeheartedly with your recommendation now ma'am. Well played. 😃

**Comment by bugfolder** [comment]
There's still a difference between placeholders (%) and wildcards (*), though; it might be useful to maintain the distinction.

* Wildcards (as they are used in URL Path visibility conditions) can appear anywhere within a path: `foo*` is a meaningful wildcard expression, but `foo%` is invalid as a layout path.
* Paths could potentially begin with a wildcard, e.g., `*` to match anything; but `%` is not allowed as a layout path.
* Placeholders can provide context. Wildcards presumably couldn't.
* Mixing wildcards and placeholders would make it pretty difficult to provide context for the placeholders: `*/%` would be pretty hard to handle as an expression.

So how about supporting them both in layout paths, but with the proviso that you can't mix `*` and `%` in the same expression?

And then maybe a "more info" link or tooltip could help explain to newcomers what the distinction is and why one would choose one over the other.

**Comment by jenlampton** [comment]
The wildcard should be able to handle the use-case for the placeholder too. 

> I'd rather Backdrop handle the context stuff in the background without people needing to understand the difference between a % (here in Layouts UI) and a * (in every other place wildcards are used). If we can do that by using a single cryptic symbol (preferably the * people already understand) instead of two, with subtle differences, that's a win :)

We might still have to support both to make this change in a way that is backwards-compatible, though I'd still prefer not to.

**Comment by bugfolder** [comment]
Ah. If we only use *, then /*/ or /* (at the end) would be interpreted as a placeholder, and any other form would be an ordinary wildcard. Reasonable?

**Comment by indigoxela** [comment]
> There's still a difference between placeholders (%) and wildcards (*), though; it might be useful to maintain the distinction.

I agree with that! The "%" placeholders mean something different (to me). Keep in mind, that % has a special meaning in hook_menu(). Think of "%node", which loads the node object. That's something totally different than "node/*" which translates to paths like "node/1", "node/2", ... "node/749".

Using "%" and "*" as equivalent might cause more confusion, a placeholder that loads an object is not the same as a wildcard on paths.

**Comment by bugfolder** [comment]
As I think about how to try to explain this to users on the "Configure layout" page, given the differences between wildcards and placeholders and the limitations on the latter, I'm more firmly thinking it will be less confusing to users to use two distinct symbols to provide the two concepts.

For example, in the "Path" section of that form, the instructions under the path might say:

> Use an asterisk ("*") for _wildcards_. Use a percent ("%") for _placeholders_. 
> 
> * A _wildcard_ * can represent any string of text and can appear anywhere in the path.
> * A _placeholder_ % can only appear between two "/" or at the end of the path and represents data that can be passed to a block as a context (see below).
> 
> [See more explanation and examples.](#)

That last link would give a popup that gives examples of both and an explanation of the difference (like the token browser). I think we do need a more extensive explanation that doesn't always take up space but is easily accessible from the help text.

The problem with using the same symbol for both arises from the limitations on placeholders, and how the Context section of the form lets you assign a context to placeholders based on their position.

Suppose we used * for both. Consider, then, the path `*/*/n*/*n/*`. Based on the current rules for placeholders, only positions 2 and 5 would be eligible to be interpreted as placeholders. If a person entered that path, but contexts were only displayed for positions 2 and 5 with no explanation, that would probably be confusing. 

But using two symbols, the proper way to enter that path and get context would be `*/%/n*/*n/%`. If a user entered a % in an invalid place, they'd get immediate feedback on why it was invalid, and it would be clearer in the "Contexts" section that the two possible contexts are for the two locations of a % symbol.

@jenlampton said:

> If we can do that by using a single cryptic symbol (preferably the * people already understand) instead of two, with subtle differences, that's a win :)

The thing is, the differences are not that subtle. It's worth using two symbols to emphasize the differences.

So I would ask folks to (re)consider keeping the % for placeholders, and adding * for wildcards.

**Comment by bugfolder** [comment]
If I may add one more bit of argument in favor of having both placeholders (`%`) and wildcards (`*`) in layout paths, versus using `*` for everything: I remember seeing that backward compatibility was an important principle in Backdrop. With that in mind, if we use two symbols, then:

* People who already used layouts with context will see everything still works the same as ever; `update.php` didn't modify their layout paths containing `%` with `*`'d versions, which would have to happen if we switch to one symbol.

* People who come newly to Backdrop and are familiar with `*` wildcard paths in in blocks (from either Backdrop or D7) will see that they can use `*` wildcards in layout paths, and so will already know what to expect from them.

* Less-experienced users won't know what a layout context is or when to use it, but don't need to worry about them until they're ready to try them out.

* More-experienced users can start mixing wildcards and placeholders whenever they're ready or feel the need.

And let me reiterate a point from my previous comment: the distinctions between wildcards and placeholders (whether we use one symbol or two) are significant enough that they need a decent amount of explanation and examples, and that explanation should be accessible directly from the UI, not tucked away in a user guide on the docs website. Ideally, right at the point where someone sees "you can can put either a wildcard or a placeholder into this path", there should be something they can click on (or tap) that takes them to enough information to help them make that choice.

**Comment by docwilmot** [comment]
The main issue with wildcards is not really to do with Layouts at all; it would require re-examining how Backdrop manages paths, altogether. 

Say you try visit the path `admin/config`. Core checks the router table to see if that path exists (ie provided by a `hook_menu()` implementation somewhere). Then **if** that path is found, core checks if a layout is assigned.

If you visit a path provided by Layout module itself, like `newpath/no-hook-menu-implementation` that could still work even if no module or core provides a `hook_menu()` implementation because Layouts creates an automatic `hook_menu()` for you and saves that in the router table too.

If we allow wildcards though: 

If you save a layout with path `admin/*` and visited `admin/config` you would succeed, because `admin/config` is already in the router table. Core first checks that path exists then finds the layout. 

BUT, if you created layout with path `newpath/*` and visited `newpath/no-hook-menu-implementation` that's impossible right now, because that's not in the router table. Core wouldn't even get as far as checking for Layouts, it would just 404.

We'd need to teach the menu system new tricks, not Layout. We'd need to tell the menu system that if a user tries to visit `newpath/no/hook/menu/implementation/at/all` (notice slashes) to calculate every possible permutation of wildcard and placeholder for this path, including:
- `newpath/no/hook/menu/implementation/at/all`
- `newpath/%/hook/menu/implementation/at/all`
- `newpath/no/hook/*/*/at/all`
- `newpath/no/hook/%/*/at/all`
- `newpath/no/*/*/%/at/all`
- `newpath/*/hook/menu/*/at/*`
etc
etc

then check the router table for every one of those. With a max path parts limit of 9 that's a lot of database work for the menu system to do, and this would need to be on every page load. Would simplify greatly if we put in some limits, like `*` can only be on the end. But this would still require a rethinking of the menu system, not just Layouts.

See [menu_get_ancestors()](https://docs.backdropcms.org/api/backdrop/core%21includes%21menu.inc/function/menu_get_ancestors/1) for how core does placeholder mapping for each path visit. We'd need to modify that function to check also for router paths with a `*` on the end for each path fragment, which wouldn't be much more in number but I cannot say what impact this would have on performance.

**Comment by bugfolder** [comment]
Ah, thanks for the very clear explanation. I'd not realized that Layouts can _create_ menu router paths, and indeed, that changes everything (or at least, it changes the basis of my assumptions).

I'm still thinking, though, that there'd be great utility in being able to apply a single layout on multiple disparate paths. I'd been thinking of layout paths as behaving like path aliases; they only applied to paths that already existed in the menu router table, so to determine layout availability, you'd filter the layout paths against the router table.

Obviously, that's not the case. But that looks like the approach that [Layout Wildcard](https://backdropcms.org/project/layout_wildcard) was following, so I'll look further at that module and its [work with path aliases](https://github.com/backdrop-contrib/layout_wildcard/issues/1) issue.

**Comment by herbdool** [comment]
I was trying Layout Wildcard and was a bit confused. I removed it to see what's possible with Layout and Views. I already understand that a creating a View page like `list/content` you can get a Layout to apply by adding a layout path of `list/content`. But I thought that maybe a layout path of `list/%` would also work. Doesn't seem to, though it does add a "string pass-through" (I should know what this does but I don't). I think Views might be unique here in that they're not path aliases, but they're also not entities like nodes or terms where Layout supports `node/%` or `taxonomy/term/%`. So maybe this would be a narrow use case which core could support? I guess it's a little complicated because Layout needs to know it's a View.

**Comment by bugfolder** [comment]
Layout's path-handling can be confusing (at least, I'm still trying to fully wrap my head around all the scenarios). Beginning with the fact that when a module and a layout both want to create the same system path (like `mypath/%`), the one that gets into the menu router table [is the one that was created first chronologically](https://github.com/backdrop/backdrop-issues/issues/4986). 

One thing I found that helps in sorting things out is to keep track of whether the path and/or layout is a _system path_ (like `node/%`) or a _normal path_ (like `node/1`). When you go to a URL like http://example.com/list/1, the normal path is `list/1`, but the system path could be either `list/1` or `list/%`, depending on what's in the menu router table.

A View can create system paths that include placeholders, if you define the View with a contextual parameter (so that the page path of the view has a % placeholder in it). Or you can create separate view displays for `list/1`, `list/2`, and so forth, but in those cases, the system paths will be `list/1` and `list/2`; both will have entries in the menu router table. If in the second case, you create a Layout with a path `list/%`, that will add an entry in the menu router table, but core Layout won't try to apply it to your View pages because core behavior matches layouts on the system path, and `list/1` != `list/%` when it comes to system path comparison.

(Some of these issues can probably be handled by upgrades in Layout Wildcard, but I'm going cautiously on implementing upgrade changes there because I don't want to do something dumb that would make it hard to change direction later.)

**Comment by herbdool** [comment]
@bugfolder your explanation helps. I tested and indeed it does work that way. A View with a path `list/%` creates a system path in menu_router table of the same. And creating a layout for `list/%` works. But a View path of `list/1` doesn't work with the same layout.

**Comment by bugfolder** [comment]
@herbdool, I've started working on the roadmap for Layout Wildcard, putting some introductory comments and explanation into [this issue](https://github.com/backdrop-contrib/layout_wildcard/issues/19). They might be helpful. The View scenario you're describing is one of the example scenarios.

**Comment by docwilmot** [comment]
Small change in new PR, allows wildcard characters at the end of layout-provided paths: So you can set a layout path as `fruit/*` and this will match `fruit/mango` and `fruit/apple`; but not `fruit/apples/fuji`.

But this doesnt match aliases, so path patterns will not match; ie `posts/*` will still not match all posts, even though they do have the URL pattern `posts/NID` because thats an alias, not a layout-provided path.

**Comment by klonos** [comment]
> ...this doesnt match aliases

Conscious decision, or technical limitation?

**Comment by docwilmot** [comment]
Conscious decision. Its a different issue (I believe there is an open issue somewhere), so different PR.

**Comment by bugfolder** [comment]
I definitely like the idea of introducing a wildcard * into layout path matching (as a concept distinct from placeholder %).

We already have a "mental model" for wildcards in paths, which are the paths used in the "URL path" visibility conditions. There, a * can stand for any set of characters, so, for example, `fruit/*` would match both `fruit/apple` and `fruit/apple/fuji`. Also, `fruit/a*` would match `fruit/apple` but not `fruit/pear`. It doesn't look like `fruit/a*` would be supported by this PR (correct?).

And URL path visibility paths can match to URL aliases, but this matching doesn't. So paths-with-wildcards match differently in the two types of path.

I don't know if it's feasible, but just in the interest of keeping the same mental model for both types of paths, should we strive to make them match the same things?

Second question: if there is a long-term goal of getting broader support for use of * wildcards in layout paths, putting this into core "locks in" this particular interpretation going forward, which makes me wonder if it might be better put into a contrib module? Sort of the way that Layout Wildcard module is (or can be viewed as) as testbed for ways of broadening the layout matching rules.

If we took * to be a true wildcard that matched any set of characters, could one achieve the same goal by letting the layout path match `fruit/*`, but then adding a URL path visibility condition to "Allow access on all pages except the following" with a path of `fruit/*/*`?

**Comment by docwilmot** [comment]
> There, a * can stand for any set of characters, so, for example, fruit/* would match both fruit/apple and fruit/apple/fuji

We could probably do that
> It doesn't look like fruit/a* would be supported by this PR (correct?) 

Correct, that wouldnt work and I think its perhaps a bit too much
> if it might be better put into a contrib module

Thats currently not possible to achieve in contrib. We'd need a way to alter `menu_get_ancestors()`; maybe an `menu_get_ancestors_alter()`, but introducing an easy way to ruin every route in your site is probably not a good idea.

Pinging @quicksketch to see what he thinks of all of this.

**Comment by docwilmot** [comment]
I've decided to complicate matters:
- updated first PR https://github.com/backdrop/backdrop/pull/3841 to allow multipart paths with a wildcard on the end. So `foo/bar/*` or `foo/*` now treated differently.
- second PR though https://github.com/backdrop/backdrop/pull/3855 provides complex wildcard paths. `foo/*/bar/*` or `foo/%/*/bar` for example.

**Comment by klonos** [comment]
Thanks @docwilmot 🙏🏼 

Can we please add inline comment(s) in `menu_get_path_wildcards()`  in the second PR, in order to explain what's going on? ...or a docblock at least.

By your description above, it sounds like the second PR is the more flexible one, so if others agree and see a use for that feature, then I say lets go with that one.

**Comment by bugfolder** [comment]
I agree that number 2 is both more flexible and closer to what people would expect how a wildcard would behave.

Once this is merged some explanation should be added to the help text of https://github.com/backdrop/backdrop-issues/issues/4868 (and this addition further underscores the utility of that help text). I suggest that as a follow-up issue, rather than binding the two issues/PRs.

**Comment by bugfolder** [comment]
Testing option 2 (PR https://github.com/backdrop/backdrop/pull/3855): I created a layout with path `*/%`, which I would have expected to be applied to paths `user/1` and `node/1`, but it's not. Is there further limitation (like "*" can't be the first element of the path)?

**Comment by bugfolder** [comment]
I've been pondering this topic for a while. What I'd envisioned (and hope to work on) is adding a new field to layouts, named something like "URL paths", that, like visibility paths, can accept any combination of text and `*`'s and can match either normal paths or URL aliases (with some exceptions, see below). The idea being that this field behaves just like the "URL paths" field of block and layout visibility conditions, so it's a behavior mostly familiar to users.

The exceptions arise due to the fact that if a layout path has a `%` placeholder that defines an object context (something more than a string pass-through), then the blocks in that layout are going to be expecting a valid context to be fed to them, and there's no guarantee that such a context could be extracted from an arbitrary URL path. So this new layout "URL paths" field won't (can't) actually be used for all paths for which the wildcard paths might match.

That, in turn, could be even more confusing than the current situation ("why does it match some `mypath*` but not others?"), so I also envisioned added some "live help text" (like we've recently added for the layout path itself) that informs the user what's going to happen with the text that they've entered. And then maybe a more/less toggle to provide even more explanation of how this field works. (Which is why I'd really like to see https://github.com/backdrop/backdrop-issues/issues/5090 get merged into core, so I can start using it for stuff like this).

Anyhow, I'll try to work on a PR as another alternative solution for folks to consider.

**Comment by bugfolder** [comment]
# Issue discussion

We have two PRs already...and as the saying goes, "now for something completely different." I've made a PR with a new approach for consideration. And, with apologies for the length, I'd like to explain the approach and why it is what it is.

Let me reiterate something from [a few comments up](https://github.com/backdrop/backdrop-issues/issues/1602#issuecomment-784809112): placeholders (%) and wildcards (*) are historically different things, with different meanings, functions, and usage rules. So I continue to see problems with trying to use one symbol for both.

But I also see that combining placeholders and wildcards in the same field (e.g., path `n*de/%/*`) would also be confusing. So I'm inclined toward the notion of having two separate fields: one for the placeholder-ish stuff, and a second field for the wildcard-ish stuff.

Stepping back a bit, I think the overall goal of this issue is to allow a single layout to be used for a wider range of request paths than the current system. And since wildcards have historically allowed matching to URL aliases, it would be nice to let layouts be used based on matches to both normal paths and URL aliases.

The normal paths part is already provided to some degree by the [Layout Wildcard module](https://backdropcms.org/project/layout_wildcard), but it would be nice to have a solution already in core, and to work with URL aliases. The thing that Layout Wildcard module does, that people seem to understand and work with, is that in addition to the (primary) Path field of the layout, there is an additional field (called "Alternative Paths" in LW) that allows matching to additional paths. So I took that as my working inspiration.

Since layouts are already pretty complicated and hard to follow, an important objective for me was to make the UI close to something that is (probably) already familiar to users: the "URL paths" field in visibility conditions for layouts and blocks (the latter was part of D7, so it would even be familiar to D7 refugees).

In visibility conditions, URL paths can match to both normal paths (`node/1`) and URL aliases (`post/your-first-post`), and so I've made that the behavior for layouts as well.

Layouts now have a new field: "URL paths", which takes a multiline text area of path patterns, containing wildcards, just like visibility conditions.

In visibility conditions, one can use wildcards ("*"), which have very different behavior and requirements from the placeholders ("%") currently used by layouts. Because wildcards and placeholders have different behaviors and different usage rules, I've chosen to keep them separate, rather than consolidating on one symbol, as some earlier comments in this issue have suggested. So that's why "URL paths" is a separate field from the existing (primary) path field in Layouts.

However, layouts have some complicating behavior that means there's not a perfect carryover from the existing concept.

First, layouts have _positioned contexts_, contexts based on arguments in particular positions in the path; some of which result in the automatic loading of objects based on the load function of the menu router path associated with the layout's path (what I'm now calling the layout's "primary path", since there's now multiple types of path inside a single layout).

The existence and use of positioned contexts means that we can't just allow a layout to be used for any request path that happens to match its URL paths. For example, suppose a layout with primary path `node/%` has a URL path that happens to match alias `accounts/myusername`, which has normal path `user/[my-uid]`, and menu router path `user/%`. The menu system, seeing `user/%`, will load the User object, and then try to pass that to the layout, which is expecting a Node object because its primary path `node/%` is based on the corresponding menu router item. Bad things would thus ensue. So I've added an additional matching rule: _a layout can only match on URL paths if the load functions for its primary path matches the load functions that the menu system is using for the request path_.

TLDR: You can't use a `node/%` layout to render a `user/%` page. (But there is a way to use one layout for both; see below.)

The whole issue of contexts is a major driver in the design of this change. A layout with path `node/%` will have a positioned context that is a Node object; it can therefore have (for example) a Relationship that creates a context of the author of the node; and perhaps many more relationships based on the assumption of the existence of that node. The possibility of a tree of derivative objects based on that context object limits what paths we could use that layout on, whatever matching rules we might have in place.

Another consideration results from the existence of layout-created pages. If we allowed layout-created pages to have URL paths, then they could be used to override existing pages, but layout-created pages have no main content block. And their whole raison d'etre is to _create_ pages, not _override_ pages. So in this implementation, URL paths are only made available to layout overrides, not to layout-created pages.

The third consideration relates to the order in which layouts are applied (which I still don't have a good UI solution for, but I think one can probably be found). Currently, since layouts are ordered underneath their primary path on the layout listing page, a user can easily specify the order in which layouts are evaluated for page rendering. But now, since any layout could potentially provide a match for the given request path, the ordering of layouts is, first, alphabetical on the menu router paths, and then within that, user-ordered. This seems unsatisfactory, and I solicit ideas on how best to let users control the ordering of layouts in light of URL path matching. (Maybe also allow user-ordering of the primary path groups?)

There needs to be some further explanation of how URL paths work in layouts, and so I have added further explanation to the "How Layouts Work" section on the layout listing page (`admin/structure/layouts`) and expandable explanation to the new "URL paths" field on the individual layout configuration page.

I also added display of the URL paths for each layout to the layout listing page so that site builders can see if URL path matching may be going on for each layout.

For some further background on the underlying concepts, see the [Wiki page for Layout Wildcard module](https://github.com/backdrop-contrib/layout_wildcard/wiki).

With that as motivation and explanation, here are some example scenarios to try out.

## View with multiple page displays

This is an example of what is probably the most common use case.

* Create a View with three page displays at paths `my-view/1`, `my-view/2`, and `my-view/3`.
* Create a Layout with primary path `my-view/1` and URL path `my-view/*`.
* Result: the layout will be used for all of the page displays of your View.

## Multiple Views that use the same layout

Another common use case, I think.

* Create several Views with different page displays.
* Create a Layout with the primary path for one of the Views and URL paths for the page paths of all the others.
* Result: the layout will be used for all of the page displays you specified.

## Use the same layout for Node and User pages?

Yes, this is possible, but requires some workaround. Note that we can't simply give a `node/%` layout a `user/*` URL path, because it will fail load function matching. The solution is to create a layout override that doesn't have any positional context objects, as follows:

* Create a View with a single page display at path `aaa` (so its layout will be checked early, because alphabetical primary paths).
* Create a Layout with primary path `aaa` (so it overrides the View just created) and URL paths `node/*` and `user/*`.
* Result: the layout will be used for both Node and User pages. (But it has neither a Node nor a User context, which is appropriate, since we expect to use it for both types of page.)
* Side effect: there is a spurious page created by the View at path `aaa`. You can set its permissions to be forbidden to hide it.

Things I'm looking for:

* General feedback: is the overall concept sound?
* Desired use cases (whether or not this implementation addresses them).
* UI improvements: better ways of presenting and/or explaining what's going on and how to configure.
* Ideas on how best to handle allowing the overall reordering of layouts with URL paths?
* Better HTML/CSS for adding the URL paths information to the layout listing page.

Tests will of course be needed, but I'll put that off until the PR survives long enough to get further along in the review process.

**Comment by indigoxela** [comment]
Pretty cool - sort of "Layout Wildcards" in core. :smirk_cat:

I played a bit in the Tugboat sandbox:

**Re View with multiple page displays / Multiple Views**

Works like a charm, is easy to understand. :+1: 

**Re Use the same layout for Node and User pages**

I can confirm that it works. The (IMO) workaround with the dummy view isn't easy to understand, though. Another limitation: without the user and node context, it's not possible to use visibility conditions (node type, user role...). At least, I don't think it is - but I'm no Layout expert. Without visibility conditions, the layout is pretty "greedy", so I'm not sure if that covers an actual use-case. But I'm curious what others can do with it.

**Comment by bugfolder** [comment]
> without the user and node context, it's not possible to use visibility conditions (node type, user role...)

Yes. But that would only be the case is someone is trying to use the same layout for both node pages (`node/%` and user pages (`user/%`), and if they are, a visibility condition based on node type would be ill-defined if they were viewing a user page.

So, the current behavior is "Even if URL paths match, don't try to use a layout for a path if they layout requires a context and the path doesn't provide it." An alternative one could consider is "If URL paths match, go ahead and use the layout and deal with the missing context." (Which then raises the question: if a visibility condition depends on the context—like node type—should it return TRUE or FALSE if the context is missing? If it returns FALSE, then that's the same outcome as "don't use the layout for this path." But if it returns TRUE, that seems to violate the basic notion of a visibility condition "only use this layout if the node context satisfies this condition."

So I think that if someone wants to use the same layout for node and user pages, the price they'll have to pay is that they can't then create visibility conditions (or relationships) that are specific to node or user pages. That seems reasonable.

Another possibility that occurred to me is to have another Layout setting, "URL path matching: loose or strict". "Strict" would be the current behavior. "Loose" would allow the layout to be used even if the path doesn't provide the right type of context, and would rely on the user's intelligent choice: basically "I'm not going to use the positioned context for anything, so it doesn't matter that it's not the right type or doesn't exist, so go ahead and use the layout if the URL path matches."

**Comment by argiepiano** [comment]
Thanks for the explanations and all the work, @bugfolder. It all makes sense to me. It'll be good to have all this in core. In the past I've used Layout Wildcard mostly to use the same layout for several different but related views. 

I haven't yet done a lot of testing, and I don't at the moment have a suggestion regarding the reordering of layouts with url paths. 

I wanted to mention that I tried testing the `aaa` path scenario you mention above, which makes `node/*` and `user/*` use that layout. I noticed that the path `node/1/edit` also used that layout (which I did not want, as I wanted the custom layout to be used for display only, not for the node edit path). After playing a bit I realized I could add a negative visibility condition for `node/*/edit` (as in `Current path is not "node/*/edit"`) and that took care of it and showed the default layout for the edit page. 

Since this change will open up the possibility of unintended use of a layout on a derivative path, would it be a good idea to add a bullet point in the "Show more" fieldset,  with something like "You can exclude derivative paths like node/*/edit by creating a URL path visibility condition".

**Comment by jenlampton** [comment]
Interesting. @argiepiano were you using an admin theme / admin layout on the edit form? Or did the new custom layout override the default administration layout?

**Comment by indigoxela** [comment]
> were you using an admin theme / admin layout on the edit form?

@jenlampton It's all in the PR sandbox. As far as I can see, all theme related settings are default.

> I noticed that the path node/1/edit also used that layout ...

@argiepiano Nice catch. The same is true for node/X/preview and node/X/delete... and the latter looks pretty odd (overridden layout, but Seven theme).

**Comment by argiepiano** [comment]
@jenlampton the new custom layout overrode `node/1/edit` as well. 

You can see that in the sandbox/tugboat instance for @bugfolder's PR. 

Open the tugboat site and visit `https://pr4132-so4oqo7zqxb2yxmoihgoqf5bedn6h4wm.tugboatqa.com/node/2/edit`

That path uses the layout  "User and Node" with the primary path `aaa`, which also contains the wildcarded URL paths `node/*` and `user/*` as explained by @bugfolder above. When you visit `node/2` this layout is used with the Basis theme. But when you visit `node/2/edit`, the same layout is used with Seven. See image below. Quite funky. When you put a negative visibility condition for the layout for not `node/*/edit` then the edit form goes back to normal. 

<img width="1215" alt="Screen Shot 2022-07-16 at 7 10 54 AM" src="https://user-images.githubusercontent.com/9938978/179356460-c8946e28-2594-455f-a23c-626e0f49dbd3.png">

**Comment by argiepiano** [comment]
One more issue (easy to fix). @bugfolder, it looks like you have to trim the content of the URL path textarea.

<img width="463" alt="Screen Shot 2022-07-16 at 7 10 12 AM" src="https://user-images.githubusercontent.com/9938978/179356556-fb5bfac4-95c0-4d63-9b3d-ef2f50a15171.png">

**Comment by bugfolder** [comment]
@argiepiano said:

> would it be a good idea to add a bullet point in the "Show more" fieldset, with something like "You can exclude derivative paths like node/*/edit by creating a URL path visibility condition".

Great idea, and a real "gotcha" candidate to confuse users. I've added a bullet to the "Show more".

> One more issue (easy to fix)...it looks like you have to trim the content of the URL path textarea.

Indeed, easy, and fixed. Thanks!

PR updated.

**Comment by quicksketch** [comment]
I tried out the PR and read the [summary](https://github.com/backdrop/backdrop-issues/issues/1602#issuecomment-1180923840) @bugfolder posted introducing it.

It seems that there's some overlap, perhaps unnecessarily, between the new `URL paths` field and the `URL path` visibility condition. I think they would both work to "reduce" the situations in which a layout could be used.

![image](https://user-images.githubusercontent.com/100206/179372965-cb4c4cfd-2082-44a2-be29-521e64f387e3.png)

Adding an additional internal path, like `node/*/edit` might expand the situations in which a layout is used, but adding an alias pattern cannot expand, only reduce the situations where it's used (is that right?)

So could we simplify down the purpose of the new field from handling both internal paths and aliases to just handle internal paths? And would that deviate from the original purpose of this issue?

**Comment by bugfolder** [comment]
> I think they would both work to "reduce" the situations in which a layout could be used.

No, they're complementary. The layout URL paths field _expands_ the situations in which a layout can be used (when you add an entry, the layout works on more paths); the visibility URL paths field _narrows_ the situations (when you add a visibility condition, it reduces the number of paths the layout could apply to).

And that's the case for both normal paths and alias paths.

In particular, the layout URL paths field lets you apply the layout to multiple unrelated menu router paths, like `foo`, `bar`, and `baz`. There's no way to do that with just visibility conditions (unless we start to allow layouts whose primary path is a wildcard, like `*`. Or a setting for layouts "instead of providing a path, use this layout on any path (make it "default-like"), and then it will be limited by visibility conditions.") (Which would be yet another way of addressing this issue.)

> So could we simplify down the purpose of the new field from handling both internal paths and aliases to just handle internal paths? And would that deviate from the original purpose of this issue?

I couldn't think of any situation where one would use an alias in a layout URL path that couldn't be reproduced by using just normal paths there and visibility conditions that involved aliases. But my motivation for including aliases was that visibility conditions on layouts and blocks based on URL paths accepted both normal paths and aliases, and I thought that to preserve the consistent "mental model," layout URL paths should, too.

**Comment by bugfolder** [comment]
> I couldn't think of any situation where one would use an alias in a layout URL path that couldn't be reproduced by using just normal paths there and visibility conditions that involved aliases

That's the behavior of the "Alternative paths" field that [Layout Wildcard](https://backdropcms.org/project/layout_wildcard) adds to the layout; it only handles normal paths, not aliases.

**Comment by quicksketch** [comment]
Sorry, I didn't write that clearly. The "URL Paths" field definitely expands the paths that the layout matches. I meant to say that when using aliases within the field, it works the same as visibility condition. Using internal paths expands the places the layout is applied, but specifying aliases will always only reduce the places it's used.

**Comment by bugfolder** [comment]
> when using aliases within the field, it works the same as visibility condition. Using internal paths expands the places the layout is applied, but specifying aliases will always only reduce the places it's used.

I'm sorry, but that's not the case, and I've put a counterexample in the Tugboat sandbox.

* I've created a View with three pages: `cat`, `dog`, and `rabbit`. 
* Then I created an alias `bunny` for `rabbit`.
* Then I created a layout with primary path `cat` and URL paths `dog` and `bunny`. 
* The layout is used on all three pages provided by the View. So in this case, adding the alias `bunny` to the URL paths field _increases_ the places it's used.

Admittedly this is a somewhat contrived example, contrived to address this particular point. I can certainly see the argument that accepting aliases in this field is unnecessary because the same outcome could be obtained by other means. But I don't see how adding an alias to that field could "reduce the places [the layout] is used."

**Comment by yorkshire-pudding** [comment]
Hi @bugfolder - l've tested this and I like it.  I've always found it weird that the layout paths were nearly all `node/%` and can see that the combination of expanding scope together with the existing visibility tools to narrow scope would allow lots of flexibility

**Comment by bugfolder** [comment]
Having seen generally positive comments about [the "URL paths" PR](https://github.com/backdrop/backdrop/pull/4132) and its approach, I'm going to move on to the missing item I mentioned above: the problem of layout ordering when URL paths are in effect. To give a concrete example, suppose we have two primary paths and three layouts:

* path **`cat`**
    * layout 1 with URL path = `foo`
* path **`dog`**
    * layout 2 with URL path = `foo`
    * layout 3 with URL path = `foo`

In the absence of URL paths, with the normal ordering of layouts, if the request path is `dog`, layout 1 won't be considered, then layout 2 will be considered, then layout 3 will be considered.

If we turn on URL paths, then if the request path is `foo`, in the current PR, the order that the layouts would be considered is, first, layout 1 (because its primary path `cat` comes first alphabetically), then layout 2, then layout 3. The problem is that there is no way for the user to specify that layout 2 or 3 should be considered firs. Layout 1 will always be considered first, because its primary path comes first alphabetically. So there needs to be a way that a user can override the alphabetical ordering of the primary paths if we're considering URL paths, where _any_ non-default layout could potentially be used.

Currently, within a primary path, the layouts can be reordered. So one could imagine letting one impose an overall global order on layouts, rather than grouping them by primary path. So, for example, if we wanted layout 3 to be considered first, we might end up with this:

* path **`dog`**
    * layout 3 with URL path = `foo`
* path **`cat`**
    * layout 1 with URL path = `foo`
* path **`dog`**
    * layout 2 with URL path = `foo`

However, breaking up the grouping by primary path seems like it would introduce confusion, since the benefits of grouping would be lost for everyone, but only the (minority of?) people who use URL paths would benefit from this global ordering.

So, the thing I'm leaning toward is adding a second field to each layout, "URL paths weight", whose default value is 0. A non-zero value of URL path weight would override the ordering, and a nonzero/non-empty value would only be allowed if the URL paths field is non-empty. That is, since only URL paths break normal ordering, only non-empty URL paths get to take advantage of URL paths weight.

So then one could do something like this:

* path **`cat`**
    * layout 1 with URL path = `foo`, URL paths weight = 0
* path **`dog`**
    * layout 2 with URL path = `foo`, URL paths weight = 0
    * layout 3 with URL path = `foo`, URL paths weight = -1

And this setting would then allow layout 3 to be considered first with request path `foo` since it has the lightest weight.

And yes, that would allow users to do self-foot-shooting by doing pathological things like

* path **`fish`**
    * layout 1 with URL path = `foo`, URL paths weight = 0
    * layout 2 with URL path = `foo`, URL paths weight = -1

where now they've just overridden an ordering by using URL path weights when they could have just reordered the layouts under the path in the usual way. We could address that with some warning on the field instructions: "Don't mess with this unless you're sure of what you're doing." I suspect the vast majority of use cases wouldn't need to use weights, since it seems very rare that someone would have multiple competing layouts with the same URL paths where ordering would matter.

The current PR displays the URL paths on the layout listing page. We'd also list the URL paths weight (if nonzero) so that (again) the user can tell most of what's going on in terms of layout path matching and ordering from what's shown on the layout listing page.

So how does that sound?

**Comment by bugfolder** [comment]
Summary of two points discussed in the Weekly Design/UX meeting:

* Regarding the order question, the use cases where someone would use the same URL path for two different menu router paths are very rare (we couldn't think of any); so we don't need to add an additional solution like "URL paths weights."

* The problem of dealing with mismatched contexts when menu router paths have placeholders (e.g., using the same layout on node/%node and user/%user) could be pretty confusing to people, and the most common use cases would be menu router paths with no contexts. So let's only allow URL paths to be added for menu paths that don't contain placeholders. Use AJAX to show/hide the "URL paths" field appropriately.

**Comment by argiepiano** [comment]
> So let's only allow URL paths to be added for menu paths that don't contain placeholders

Well, one of the most common use cases for me is a views path that has arguments (I use String passthrough in the Layout). I was looking forward to this being in core, since I want to use the same layout for several related views that use the same arguments but have different URLs. This decision essentially precludes this use case (which I want to believe is pretty common?).

**Comment by bugfolder** [comment]
Hmm, that use case is good to know about, so maybe we should reconsider that decision. Or think about finding an intermediate case between "no placeholders" and "placeholders that result in specific object contexts." Certainly string pass-throughs would be pretty innocuous. 

That comes back to one of the ideas I mentioned in the meeting: have a setting in the layout that says "allow placeholders, and I promise to configure URL paths so that the layout will only be used on paths that have placeholders in the right place."

Meanwhile, I've updated the PR per the above, and also made some UX changes (improvements, I hope), which I'll lay out in the next comment.

**Comment by bugfolder** [comment]
In the new PR, here's what the layout listing looks like. I had originally just added the URL paths as an item list to the layout description, but here I've used a sentence so that the language on the layout mirrors the language used for visiblity conditions.
![layout_listing](https://user-images.githubusercontent.com/29214084/181663797-2d8ec827-816f-4922-96ae-4b9f107be63f.jpg)

Here's what the URL paths field looks like for override paths without placeholders (i.e., it is enabled). Note the expandable "Show more/Show less" section, which I think is important to help users understand how this works. It's normally collapsed, but shown here expanded.
![override_path](https://user-images.githubusercontent.com/29214084/181663880-4eaddf8c-f255-4759-a9ea-9e754e5e5e50.jpg)

And here's what it looks like when the field is disabled. I show it disabled, rather than fully hidden, so that users thinking "hey, where's the URL paths field?" would know why they can't use it.
![placeholder_path](https://user-images.githubusercontent.com/29214084/181663973-8b8d43c7-19ee-4abd-a8be-a100157d97f6.jpg)

**Comment by docwilmot** [comment]
> So, the thing I'm leaning toward is adding a second field to each layout, "URL paths weight"

This scenario, comment https://github.com/backdrop/backdrop-issues/issues/1602#issuecomment-1193032291, I would imagine should be quite unusual. I imagine if you really want path `foo` to use layout 2, you should create a layout with primary path `foo`.

**Comment by bugfolder** [comment]
> This scenario, comment https://github.com/backdrop/backdrop-issues/issues/1602#issuecomment-1193032291, I would imagine should be quite unusual.

Yes, I'm dropping the "url path weights" idea for that reason. Thanks for the F/B!

I do have a new iteration in progress, hope to have something to post fairly soon.

**Comment by bugfolder** [comment]
The scenario that @argiepiano describes [above](https://github.com/backdrop/backdrop-issues/issues/1602#issuecomment-1198722756) sounds common enough that it seems like it should be accommodated "out of the box" with this PR. So I've restored URL path matching for menu router paths that have placeholders, but enforce that the load functions (if any) must match between the router item of the layout path and the router item of the request path.

As I was trying this out for various scenarios, I found that a not-uncommon hack for making URL paths useful for paths with incompatible load functions (like `node/%` and `user/%`) was creating a "dummy" page whose path was used as the primary path for the layout (even though the page at the dummy path was never actually loaded). This seemed useful enough that I've made it an official part of the module, and I'm calling it the "layout null path". (Certainly open to other names.)

As an example of its utility, if you create a layout with primary path `null` and URL paths `node/*` and `user/*`, you get a layout that applies to both node pages and user pages. (The null path value (`null`) is configurable.)

So URL paths make layouts pretty powerful, but as much as we strive to make things discoverable from within the admin website UI (and I've added a lot of conditional explanation there), there really needs to be some extensive documentation for the whole story. So I've written a documentation page, intended to go on the docs site (which would be linked from the website UI). For now, I'm attaching it to this comment. Please read through it. It includes examples of several scenarios that I think would be common use cases. (If you can think of other use cases that should be addressed and/or explained, please add them.)

There is the beginnings of an automated test helper module, "Layout URL Paths Tests", which is (temporarily) not hidden; if you enable it, you can see the four examples described in the documentation.

The PR is updated (and rebased); please give it a whirl.

[Documentation of URL Paths.md](https://github.com/backdrop/backdrop-issues/files/9227763/Documentation.of.URL.Paths.md)

**Comment by jenlampton** [comment]
> Well, one of the most common use cases for me is a views path that has arguments (I use String passthrough in the Layout). I was looking forward to this being in core, since I want to use the same layout for several related views that use the same arguments but have different URLs. This decision essentially precludes this use case (which I want to believe is pretty common?).

> So I've restored URL path matching for menu router paths that have placeholders,

I don't understand how these two things are related. The menu router paths for these views do not have placeholders. What am I missing?

**Comment by bugfolder** [comment]
> The menu router paths for these views page displays [a views path that has arguments] do not have placeholders. What am I missing?

That those menu router paths do indeed have placeholders. Try this.

Create a View with a contextual filter whose page path is `myview/%`.
Go look at the `{menu_router}` table. There will be an entry whose `path` field is `myview/%`.

**Comment by bugfolder** [comment]
Note from today's meeting: it would be nice to give instant feedback, or instant in-place correction, if someone tries to use a "*" in the layout's primary path (or, for that matter, tries to use a "%" in the URL paths field).

Another comment from today's meeting: a suggestion that a UI akin to Context's popup dialog for adding additional paths one-at-a-time (rather than a textarea field like URL paths in visibility conditions) might be preferable. (H/T @quicksketch.)

**Comment by jenlampton** [comment]
> Create a View with a contextual filter whose page path is myview/%

I don't think I've ever done this. Do you just put the percent symbol in the "Path" field on the page display? (and also, why would you ever do this? Is there an 80% use-case?) 

To test, I created a normal view with a contextual filter (the path I used was "foo" but did not contain a %). I checked the  `{menu_router}` table to be sure views wasn't adding a placeholder because of the contextual filter, and it was not. 

> The use case that @argiepiano describes is the case "A dynamic View with multiple page displays" described in the documentation document.

Yes, I saw that and added it into the description of this ticket also. But as I said, there were no placeholders in the `{menu_router}` table after I tested that use-case, hence my question.

**Comment by yorkshire-pudding** [comment]
Hi @jenlampton it's on the site that you worked on before me. It uses it for different taxonomy filters

**Comment by jenlampton** [comment]
@yorkshire-pudding But adding a % at the end of the path is completely unnecessary... so why do people do it? Is it a mistake, or is there a use-case?

**Comment by bugfolder** [comment]
> I don't think I've ever done this. Do you just put the percent symbol in the "Path" field on the page display? (and also, why would you ever do this?

I've done it for years in D7, and it's on my main live site. For example, I have a node type of "Class", which has fields of "Year" and "Season", and there is a View that provides a listing of classes at, for example, path `classes/2022/3` (for all classes in the 3rd season of 2022).

And yes, you put a '%' in the path of the menu item for each of the Views contextual filter arguments.

> Is this an 80% use-case?

Well, it's all over Views in core. Do a search for `views_handler_argument_` through core, and you'll find a whole slew of handlers that are intended to work with exactly this type of path.

I also solicit @argiepiano to say more about his use case(s).

**Comment by docwilmot** [comment]
Does this PR actually enable wildcards (*) now? I dont see how it would. And its not working in the sandbox.

**Comment by argiepiano** [comment]
Sorry, I've been busy and not paying much attention here. It's true that you don't really need placeholders in Views paths when you are dealing with paths that contain contextual filter(s) as the last parts of the path (placeholders not needed in the views path for `my-path/%` or `my-path/%/%`), but you do need them when you are creating multiple paths where the last part is a string, for example `my-path/%/csv` or 'my-path/%/rss`. And those will create menu router items with placeholders, thus my concern.

**Comment by bugfolder** [comment]
> Does this PR actually enable wildcards (*) now? I dont see how it would.  And its not working in the sandbox.

It does, and it's working in the sandbox (at least, for the documented cases). I enabled the test module that creates the use cases referred to in the documentation.

I see you (or someone) has created a layout-created page that overrides another layout-created page (and so they're both listed as layout-created pages in the listing). That's a use case I hadn't considered, and that we briefly touched on the design meeting. I think the consensus there is that this is a case that ought to be supported, and so the PR will need to be updated to handle that case.

And also the case of a View with contextual filters that don't include placeholders. (After all these years, I'd just assumed you needed the placeholders and so had always used them.)

**Comment by bugfolder** [comment]
> I think the consensus there is that [a layout-created page overriding another layout-created page] is a case that ought to be supported, and so the PR will need to be updated to handle that case.

And in pondering how to handle this...the reason I didn't originally allow layout-created pages (LCPs) to be used to override other non-layout-created pages via URL paths was that the former don't have the "Main content" block, and I thought it would be really confusing for a user to create a LCP, give it a URL path of `node/*`, and then when they visited `node/1`, they'd see their layout but no node.

So maybe the right behavior is that LCPs with URL paths can only override other LCPs. But then that means more quirks that need to get documented in the documentation.

(Maybe it's best instead to stuff all the functionality into Layout Wildcard module, and let the resolution of _this_ issue go back to its original title: do instant-feedback/correction if a user tries to use a `*` instead of a `%` in the primary path of the layout.)

**Comment by docwilmot** [comment]
This issue has become so complex because we're trying to add:
- wildcard support
- multiple layout paths
- layout path alias support

I believe the last two tasks had previous issues somewhere. This issue started for the first task. Maybe we should tackle fewer than 3 at a time? If this gets bogged down I'd recommend we tackle the 2nd task first (I think that might be the more complex UI and API task), then come back to this one.
